relayer-engine/src/walletMonitor/balances.ts

```diff
@@ -0,0 +1,531 @@
+
+  for (const privateKey of privateKeys) {
+    balancePromises.push(pullEVMNativeBalance(chainConfig, privateKey, provider));
+    pullEVMTokens(privateKey, chainConfig, tokensToMonitor, provider);
+  }
+
+  return balancePromises;
+}
+
+export async function pullAllSolanaBalances(chainInfo: ChainConfigInfo, privateKeys: string[]): Promise<WalletBalance[]> {
+  const balances: WalletBalance[] = [];
+
+  for (const privateKey of privateKeys) {
+    const UintPK = Uint8Array.from(JSON.parse(privateKey));
+    const [nativeBalances, tokenBalances] = await Promise.all([
+      pullSolanaNativeBalance(chainInfo, UintPK),
+      pullSolanaTokenBalances(chainInfo, UintPK),
+    ]);
+
+    if (nativeBalances) balances.push(...nativeBalances);
+    if (tokenBalances) balances.push(...tokenBalances);
+  }
+
+  return balances;
+};
+
+async function pullTerraBalance(
+  lcd: LCDClient,
+  walletAddress: string,
+  tokenAddress: string,
+  chainId: TerraChainId
+): Promise<WalletBalance> {
+  const tokenInfo: any = await lcd.wasm.contractQuery(tokenAddress, {
+    token_info: {},
+  });
+  const balanceInfo: any = await lcd.wasm.contractQuery(tokenAddress, {
+    balance: {
+      address: walletAddress,
+    },
+  });
+
+  if (!tokenInfo) throw new Error('Token info not found');
+  if (!balanceInfo) throw new Error('Balance info not found');
+
+  return {
+    chainId,
+    balanceAbs: balanceInfo?.balance?.toString() || "0",
+    balanceFormatted: formatUnits(
+      balanceInfo?.balance?.toString() || "0",
+      tokenInfo.decimals
+    ),
+    currencyName: tokenInfo.symbol,
+    currencyAddressNative: tokenAddress,
+    isNative: false,
+    walletAddress: walletAddress,
+  };
+}
+
+async function pullSolanaTokenBalances(
+  chainInfo: ChainConfigInfo,
+  privateKey: Uint8Array
+): Promise<WalletBalance[]> {
+  const keyPair = Keypair.fromSecretKey(privateKey);
+  const connection = new Connection(chainInfo.nodeUrl);
+  const _logger = getScopedLogger(["pullSolanaTokenBalances"], logger());
+
+  let allAccounts
+  try {
+    allAccounts = await connection.getParsedTokenAccountsByOwner(
+      keyPair.publicKey,
+      { programId: TOKEN_PROGRAM_ID },
+      "confirmed"
+    );
+  } catch (error) {
+    _logger.error(`Error token balances for public-key: ${keyPair.publicKey}. ${error}`);
+    return [];
+  }
+
+  const output: WalletBalance[] = [];
+
+  for (const account of allAccounts.value) {
+    const mintAddress: string[] = [account.account.data.parsed?.info?.mint];
+
+    if (!mintAddress) {
+      _logger.warn(`No mint address found for public-key: ${keyPair.publicKey}. Skipping...`);
+      continue;
+    }
+
+    try {
+      const mdArray = await getMetaplexData(mintAddress, chainInfo);
+      let cName: string = "";
+      if (mdArray && mdArray[0] && mdArray[0].data && mdArray[0].data.symbol) {
+        const encoded = mdArray[0].data.symbol;
+        cName = encodeURIComponent(encoded);
+        cName = cName.replace(/%/g, "_");
+      }
```

Comment:
We don't need the mint of the account, the token balance should be available in the parsed data of the account itself in the `.amount` field (see [here](https://github.com/solana-labs/solana-program-library/blob/0e48b25a8f9175d67825f061a9913cddff1837c9/token/js/src/state/account.ts#L17))